commands/check.go

```diff
@@ -107,30 +86,202 @@ func (c *check) Execute() error {
+					continue
+				}
+
+				if !strings.HasPrefix(pre, "alpha") &&
+					!strings.HasPrefix(pre, "beta") &&
+					!strings.HasPrefix(pre, "rc") {
+					// additional variant, not a prerelease segment
+					continue
+				}
+			}
+
+			if cursorVer != nil && (cursorVer.GreaterThan(ver) || cursorVer.Equal(ver)) {
+				// optimization: don't bother fetching digests for lesser (or equal but
+				// less specific, i.e. 6.3 vs 6.3.0) version tags
+				continue
+			}
+		}
+
+		tagRef := repo.Tag(identifier)
+
+		digest, found, err := headOrGet(tagRef, opts...)
+		if err != nil {
+			return resource.CheckResponse{}, fmt.Errorf("get tag digest: %w", err)
+		}
+
+		if !found {
+			continue
+		}
+
+		tagDigests[identifier] = digest.String()
+
+		if ver != nil {
+			versionTags[ver] = tagRef
+
+			existing, found := digestVersions[digest.String()]
+
+			shouldSet := !found
+			if found {
+				if existing.Prerelease() == "" && ver.Prerelease() != "" {
+					// favor final version over prereleases
+					shouldSet = false
+				} else if existing.Prerelease() != "" && ver.Prerelease() == "" {
+					// favor final version over prereleases
+					shouldSet = true
+				} else if strings.Count(ver.Original(), ".") > strings.Count(existing.Original(), ".") {
+					// favor more specific semver tag (i.e. 3.2.1 over 3.2, 1.0.0-rc.2 over 1.0.0-rc)
+					shouldSet = true
+				}
+			}
+
+			if shouldSet {
+				digestVersions[digest.String()] = ver
+			}
+		}
+
+		if from != nil && identifier == from.Tag && digest.String() == from.Digest {
+			// if the 'from' version exists and has the same digest, treat its
+			// version as a cursor in the tags, only considering newer versions
+			//
+			// note: the 'from' version will always be the first one hit by this loop
+			cursorVer = ver
+		}
+	}
+
+	var tagVersions TagVersions
+	for digest, version := range digestVersions {
+		tagVersions = append(tagVersions, TagVersion{
+			TagName: versionTags[version].TagStr(),
+			Digest:  digest,
+			Version: version,
+		})
+	}
+
+	sort.Sort(tagVersions)
+
+	response := resource.CheckResponse{}
+
+	for _, ver := range tagVersions {
+		response = append(response, resource.Version{
+			Tag:    ver.TagName,
+			Digest: ver.Digest,
+		})
+	}
+
+	if latestTag != "" {
+		digest := tagDigests[latestTag]
+
+		_, existsAsSemver := digestVersions[digest]
+		if !existsAsSemver {
+			response = append(response, resource.Version{
+				Tag:    latestTag,
+				Digest: digest,
+			})
+		}
+	}
+
+	return response, nil
+}
+
+type TagVersion struct {
+	TagName string
+	Digest  string
+	Version *semver.Version
+}
+
+type TagVersions []TagVersion
+
+func (vs TagVersions) Len() int           { return len(vs) }
+func (vs TagVersions) Less(i, j int) bool { return vs[i].Version.LessThan(vs[j].Version) }
+func (vs TagVersions) Swap(i, j int)      { vs[i], vs[j] = vs[j], vs[i] }
+
+func checkTag(tag name.Tag, source resource.Source, version *resource.Version, opts ...remote.Option) (resource.CheckResponse, error) {
```

Comment:
GitHub won't let me comment on the line I want to, but:

I think we need to put the `tag` in the version here too, since now `in` populates `./tag` based on the version

Reply:
The following pipeline demonstrates this:

```yaml
resource_types:
- name: registry-image-pr
  type: registry-image
  source:
    repository: concourse/registry-image-resource
    tag: pr-alpine-214

resources:
- name: nginx
  type: registry-image-pr
  source:
    repository: nginx
    tag: alpine

jobs:
- name: job
  plan:
  - get: nginx
  - task: task
    config:
      platform: linux
      image_resource:
        type: registry-image
        source: {repository: busybox}
      inputs:
      - name: nginx
      run:
        path: cat
        args: [./nginx/tag]
```

it also looks weird in the UI, as it has an empty tag in metadata

Reply:
fixed!